Implement downloading archived item + QA runs as multi-WACZ

tw4l opened the pull request:
Fixes #1412 

## Changes

### Backend
- Adds `all-crawls`, `crawls`, and `uploads` API endpoints to download archived item as multi-WACZ
- Adds backend tests for new endpoints

### Frontend

Adds ability to download archived item from:

- Button in archived item detail Files tab
- Archived item details actions menu
- Archived items list menu

## Screenshots

### Files tab

<img width="1183" alt="Screenshot 2024-07-16 at 1 38 12 PM" src="https://github.com/user-attachments/assets/aa800383-14b4-4466-9913-99d3ec9f0277">

### Detail actions menu

<img width="1184" alt="Screenshot 2024-07-16 at 1 40 06 PM" src="https://github.com/user-attachments/assets/7bf7c49b-f667-4b6c-811e-3727dd842bbf">

### Archived items list menu

<img width="1181" alt="Screenshot 2024-07-16 at 1 37 32 PM" src="https://github.com/user-attachments/assets/54e3a436-ea91-4769-838f-fd0ada8e5991">

tw4l committed:
Add backend endpoint to download item as multi-wacz

tw4l committed:
Add download link to archived item actions menu

tw4l committed:
WIP: Add download link to archived items list menu

Not working yet - doesn't trigger download

tw4l committed:
Add Download button to files tab

tw4l committed:
Check all-crawls download at same time as crawls

tw4l committed:
Fix pytest parametrize decorator

tw4l committed:
Fix downloading from archived items list menu

This commit removes some safety checks on the btrix-dropdown
containing the menu items. In Brave, Chrome, and Firefox, the
behavior that the checks were supposed to guard against aren't
happening after the change.

tw4l committed:
Generalize language in detail action menu

emma-sg reviewed (approved):
Looks good to me! We could stand to remove the other `preventDefault` calls on click events on overflow dropdowns in list views too, since I think now that we've migrated to the new tables we're using links that don't cover the dropdown menu any more, but that could be a cleanup item I suppose.

Shrinks99 commented on frontend/src/pages/org/archived-item-detail/archived-item-detail.ts:
```diff
@@ -275,7 +275,21 @@ export class ArchivedItemDetail extends TailwindElement {
+                variant="primary"
+              >
+                <sl-icon slot="prefix" name="cloud-download"></sl-icon>
+                ${msg("Download")}
```
Would _probably_ like to keep this labeled the same for consistency?  Also considered "Download all" but that makes it sound like they'll all be started individually.  Tooltip is good!

```suggestion
                ${msg("Download Item")}
```

Shrinks99 commented on frontend/src/pages/org/archived-item-detail/archived-item-detail.ts:
```diff
@@ -618,7 +647,7 @@ export class ArchivedItemDetail extends TailwindElement {
               >
                 <sl-icon name="trash3" slot="prefix"></sl-icon>
-                ${msg("Delete Crawl")}
+                ${msg("Delete Item")}
```
Good catch!

tw4l committed:
Update frontend/src/pages/org/archived-item-detail/archived-item-detail.ts

Co-authored-by: Henry Wilkinson <henry@wilkinson.graphics>

tw4l commented:
> Looks good to me! We could stand to remove the other `preventDefault` calls on click events on overflow dropdowns in list views too, since I think now that we've migrated to the new tables we're using links that don't cover the dropdown menu any more, but that could be a cleanup item I suppose.

Yeah I think maybe that's best as a separate cleanup task? Thank you!

tw4l committed:
Merge branch 'main' into issue-1412-download-archived-item-btn

SuaYoo commented on frontend/src/pages/org/archived-item-detail/archived-item-detail.ts:
```diff
@@ -275,7 +275,21 @@ export class ArchivedItemDetail extends TailwindElement {
+            <sl-tooltip content=${msg("Download all files as a single WACZ")}>
+              <sl-button
+                href=${`/api/orgs/${this.orgId}/all-crawls/${this.crawlId}/download?auth_bearer=${authToken}`}
+                download}
```
Typo?
```suggestion
                download
```

tw4l committed:
Update frontend/src/pages/org/archived-item-detail/archived-item-detail.ts

Co-authored-by: sua yoo <sua@webrecorder.org>

tw4l commented:
@SuaYoo thanks for catching the typo!

ikreymer committed:
Merge branch 'main' into issue-1412-download-archived-item-btn

ikreymer committed:
- add single-WACZ download endpoint for QA runs
- add test for endpoint
- use new endpoint in frontend

ikreymer committed:
add missing import

ikreymer committed:
better filenames for qa run downloads

ikreymer commented:
Also added download of QA run data as multi-WACZ as well, as now we are downloading only the first WACZ

ikreymer committed:
more missing import fix

tw4l commented:
> Also added download of QA run data as multi-WACZ as well, as now we are downloading only the first WACZ

Thanks for this!

tw4l commented:
@SuaYoo mind taking another look at this when you have a moment?

ikreymer committed:
Merge branch 'main' into issue-1412-download-archived-item-btn

ikreymer reviewed (approved):
UI and other changes look good - (and hopefully the QA downloads s well).

Investigating potential crc32 mismatches in the generated WACZs, though I think that may have to do with the streaming zip library we're using and not related to the changes in this PR itself.

Shrinks99 reviewed (approved):
Tested the UI, looks good, haven't tested the feature myself.  Other issues are best dealt with by other folks, approving my end of things! :)

ikreymer committed:
fix streaming downloads with latest 'stream-zip':
- update to a new version 'stream-zip' which does not require crc32 to produce a stream WACZ
- don't specify crc32 in streaming WACZ as it may be incorrect or missing, compute on the fly
- update types for latest 'stream-zip'

ikreymer commented:
> Investigating potential crc32 mismatches in the generated WACZs, though I think that may have to do with the streaming zip library we're using and not related to the changes in this PR itself.

I believe this should now be fixed, with a new custom fork of stream-zip! Crawler was actually computing crc32 incorrectly, with latest fix, not requiring crc32 at all

ikreymer commented:
One other thing to consider: what do we want the default names of the WACZs to be? Should it just be the crawl id, or perhaps the workflow name to be more user-friendly? (Can also revisit later)

tw4l commented:
> One other thing to consider: what do we want the default names of the WACZs to be? Should it just be the crawl id, or perhaps the workflow name to be more user-friendly? (Can also revisit later)

I think crawl id is okay for now? It's similar to the WACZ filenames in the Files list (with prefix and without the instance suffix). With the workflow name it might get a bit long?
